src/v/storage/key_offset_map.h

```diff
@@ -91,4 +93,81 @@ class simple_key_offset_map final : public key_offset_map {
     size_t _max_keys;
 };
 
+/**
+ * A key_offset_map in which the key space is mapped to sha256(key).
+ *
+ * This container does not auto-grow on insert, and a default initialized
+ * instance has zero capacity. To add capacity call `reset(size_bytes)`. This is
+ * futurized to avoid reactor stalls when allocating a large amount of memory
+ * for the container.
+ */
+class hash_key_offset_map : public key_offset_map {
+public:
+    seastar::future<std::optional<model::offset>>
+    get(const compaction_key& key) const override;
+
+    seastar::future<bool>
+    put(const compaction_key& key, model::offset offset) override;
+
+    model::offset max_offset() const override;
+
+    size_t size() const override;
+    size_t capacity() const override;
+
+    /**
+     * Reset the map to have capacity \p size bytes. A call to reset() will
+     * destroy all existing data in the map and reset the size to zero.
+     */
+    seastar::future<> reset(size_t size);
+
+    /**
+     * The ratio of hash table searches (e.g. one get or put) to the number of
+     * underlying accesses made into the hash table.
+     */
+    double hit_rate() const;
+
+private:
+    using hash_type = hash_sha256;
+
+    /**
+     * hash table entry.
+     */
+    struct entry {
+        hash_type::digest_type digest{};
+        model::offset offset;
+        bool empty() const;
+    };
+
+    /**
+     * Uses successive chunks of sizeof(index_type) bytes taken from hash(key)
+     * as probes into the hash table. When `next()` returns null then the caller
+     * should switch to linear probing.
+     */
+    struct probe {
+        using index_type = uint32_t;
+        static_assert(sizeof(index_type) <= hash_type::digest_size);
+
+        explicit probe(const hash_type::digest_type&);
+
+        std::optional<index_type> next();
+
+        hash_type::digest_type::const_pointer iter;
+        hash_type::digest_type::const_pointer end;
+    };
```

Comment:
So the idea here is to probe at random places in the backing vector. This is neat. Does this smartness have a name?

Reply:
I looked extensively to see if this had a name, but I don't think it does.

IMO this is probably just a different flavor/generalization of _double hashing_ where if the first hash position is a collision a different hash function is used as the next probe. Here the hash output is large enough that we wouldn't use all those bits anyway for the first position.